Bilinear/nearest-neighbour  regridding with land-sea mask

zfan001 opened the pull request:
Description
New Code for Regridding with Land-Sea Mask [Author: Zhiliang(Julian) Fan & Tom Gale]

(1)	Add new finite-element-based bilinear regridding with land-sea mask awareness. This method is more suitable for smooth continuous variables than the nearest neighbour regridding method. The implementation is very fast (only 10-15% increase on computational time as compared with the bilinear regridding without considering land-sea mask). 
(2)	Add new implementation of nearest neighbour regridding with land-sea mask awareness. This new implementation achieved a significant speed-up as compared with the original implementation. 
(3)	Add new implementation of  nearest-neighbour regridding, which is much faster than the original Iris version.
(4)    Add new implementation of  bilinear regridding, which achieved same results and similar computational time as the Iris version.
(5)   A new directory ./improver/regrid is created for regridding. Old regridding code (in ./improver/standardise.py)and new regridding code are moved into this directory, and re-organized into several files. A unit testing directory ./improver_tests/regrid also created accordingly.
(6) The implementation is extended from the existing RegridLanSea options("bilinear","nearest","nearest-with-mask"). New options are "nearest-2","bilinear-2","nearest-with-mask-2","bilinear-with-mask-2" (see ../regrid/landsea.py)
(7)  New unit test is added  by a careful-designed regridding example, in which so many different scenarios and combinations turn up (passed). 
(8)  New acceptance test cases are also added for four new options respectively.
(9) The code assumes that the source grid is in rectilinear latitude/longitude coordinate system, and the target grid could be in rectilinear latitude/longitude coordinate system or  Lambert Azimuthal Equal Area coordinate system ( a curvilinear coordinate system). 
(10) Five Jupyter notebooks are available for this new feature, which will be added into improver-aux. 
Testing:
 - [X] Ran tests and they passed OK
 - [X] Added new tests for the new feature(s)

CLA
 - [X] If a new developer, signed up to CLA

MoseleyS commented:
Thanks @zfan001 for this PR. (cc @tjtg). This is the largest PR for a while and it will take a bit of reviewing. What is the time scale for getting this reviewed? Is it holding any other work up?

zfan001 commented:
Hi Stephen,
Yes, a lot of new code in this PR, and the code is quite complicated as well.   The code is not holding up other work.  I will fix the codecov patch problem ASAP.  BTW, it adds several new options for regridding. It should have no impact if new options are not used.   Kind regards, Julian

zfan001 commented:
I re-designed the unit test case to ensure code coverage. I also add a test for coordinate system conversion.
codecov test is okay now.

Kind regards, Julian

dmentipl commented on improver/regrid/landsea.py:
```diff
@@ -0,0 +1,454 @@
+        self.input_land = None
+        self.nearest_cube = None
+        self.output_land = None
+        self.output_cube = None
```
It seems to me that the attributes `nearest_cube`, `output_cube`, `input_land`, and `output_land` should just be variables passed in/out of `correct_where_input_true `as opposed to class attributes

zfan001 replied:
Done as suggested

dmentipl commented on improver/regrid/landsea2.py:
```diff
@@ -0,0 +1,208 @@
+        weights = np.zeros((out_latlons.shape[0], 4), dtype=np.float32)
+        # Fill in distances and weights
+        if NEAREST in self.regrid_mode:
+            for i in range(4):
```
Perhaps here the "4" could be a global variable to avoid ["magic numbers"](https://en.wikipedia.org/wiki/Magic_number_(programming)).

zfan001 replied:
added NUM_NEIGHBOURS = 4

dmentipl reviewed (commented):
Thanks @zfan001 for this work. This is a large PR and I don't understand all of the technical details. My review is focussed on software and code style.

A few general comments:

- There seems to be a missing `__init__.py` file in regrid.
- The docstrings could be better formatted. For example, sentences are usually capitalised with periods at the end. There are quite a few double spaces. (These can be found, e.g., in Vim with this regex `[^\s] [^\s#]`. The hash is there to ignore inline comments formatted by Black.) See these resources too: [PEP 257](https://www.python.org/dev/peps/pep-0257/) and [Google Python style guide](https://google.github.io/styleguide/pyguide.html#s3.8.1-comments-in-doc-strings).
- I may have made some suggestions/comments on existing code that was moved in the PR. Feel free to ignore those if it's not relevant to this PR.
- The unit tests are mostly done via the cli interface. E.g. there are no unit tests of the functions in bilinear.py, idw.py, landsea2.py, or nearest.py. It might be good to add a few unit tests for these functions directly.

owena11 commented on improver_tests/regrid/test_RegridWithLandSeaMask.py:
```diff
@@ -0,0 +1,418 @@
+    create a simple cube from data and lat/lon coords
+    args:
+        cube_array (numpy.ndarray):
+            data value (two dimensional)
```
Ever going to be applied to multi-level data? Another case that might be useful to have test coverage for if it is needed.

zfan001 replied:
Acceptance test used multi-level data, and Jupyter notebooks did multi-level as well.

owena11 reviewed (changes requested):
Hey Julian,

Tried to cover off as much detail as possible but sure I've missed something given the size of the PR, but it all looks generally good! I've mainly focused on the Bilinear regridder with only a quick check of the nearest regridder, but the review was growing in size... thought it was worth putting out these comments first before checking Nearest in more detail.  I know @fionaRust is also taking a look.

Summary:
 - Bilinear regridder checks out and is as described 👍 
 -  Understanding all assumptions about the input data is the most important thing missing from my point of view, maybe add in some optional validation of the data? (Think there is a hidden assumption about all coordinates being ascending and a stated assumption regarding the equal spacing of the source).
 - A bit of refinement of the structure of the code added would help massively making it approachable for some looking at this in the future and maintaining it.
 - A few area's where I wasn't clear about the intended behaviour, particularly in the edge cases of surface miss matches, I've just asked questions around, as I don't know if it's really right or wrong.

fionaRust commented:
Hi @zfan001 

I've started review this today and I'll continue tomorrow but I have a couple of questions in the meantime:
* You mention in your introduction that there are some jupyter notebooks to go with this PR that you want to put into improver_aux. Have you already made this PR?
* Can you or @tjtg send me an email to let me know how best to access a copy of your new acceptance test data?

Thanks

tjtg commented on improver/regrid/idw.py:
```diff
@@ -0,0 +1,219 @@
+    # Convert output to XYZ and query the KDtree for nearby input points
+    out_x, out_y, out_z = ecef_coords(out_latlons[:, 0].flat, out_latlons[:, 1].flat)
+    out_coords = np.c_[out_x, out_y, out_z]
+    distances, indexes = in_kdtree.query(out_coords, k)
```
IMPROVER code is generally set up so that each task corresponds to a single CPU thread and parallelism is handled at the suite level.
I'd suggest avoiding multiple workers if possible for simplicity.
If there's an identified performance bottleneck here, the `workers` option could be handy, but doing that would need special handling for this task in the suite.

zfan001 commented:
Hi Daniel, Aled & Fiona, @dmentipl @owena11 @fionaRust 

Thank you very much for your review.   I came back to work this Monday after a long break. I am working on some code changes based on your feedback, and  will reply to each item soon.

@Fiona,  Jupyter Notebooks and acceptance test data will be ready tomorrow. I will submit them for merging soon. Tom will help to make the files accessible to you. 

Kind regards,
Julian

fionaRust commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,401 @@
+    lon_d = lon_d_1 if lon_d_1 > lon_d_2 else lon_d_2
+
+    domain = iris.Constraint(
+        latitude=lambda val: lat_min - 2.0 * lat_d < val < lat_max + 2.0 * lat_d
```
What the reason for choosing a domain this big to cut out? I agree you need a bit of a border on it, but twice the size of the domain seems quite large.

 Also shouldn't cube_in and cube_in_mask always be the same shape and size? We don't currently check this but from I think this should always be true, can you think of a reason why it wouldn't be true. I wonder if we did check that they are on the same domain much higher up in the code we could simplify some of your code. 

We also have two different utilities that can both be used to extract subsets of cubes. I'm not sure if either could replace this code, I'll have a think. (The two utilities are: https://github.com/metoppv/improver/blob/3cfbf3323d3a693a9c61fec13350295b85d03676/improver/utilities/cube_extraction.py#L228 and https://github.com/metoppv/improver/blob/3cfbf3323d3a693a9c61fec13350295b85d03676/improver/utilities/cube_extraction.py#L261

zfan001 replied:
The first code is to add just one grid size, but Iris slice failed one test case (not worked as I expected). I  changed it to twice.
 
cube_in and  cube_in_mask could be different in old BOM's code, and therefore we keep this flexible way.

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+    masked_distances += np.finfo(np.float32).eps
+    inv_distances = 1.0 / masked_distances
+    # add power 1.80 for inverse diatance weight
+    optimum_power = 1.80
```
Done as suggested

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+
+    # Process groups of output points which have each one of the four surrounding
+    # input points as mismatched surface type
+    for i in range(4):
```
Done as suggested

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+
+    # Cases with one mismatched input point by adjusting bilinear weights
+    # leftover_bilinear is special cases of using inverse distance weighting
+    one_mismatch = np.where((count_same_surface_type == 3))[0]
```
Done as suggested

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+
+    Args:
+        in_latlons(numpy.ndarray):
+            tource points's latitude-longitudes
```
fixed

zfan001 commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,401 @@
+        List[str]:
+            List of coordinate names
+    """
+    cube_coord_names = []
```
Done as suggested

zfan001 commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,401 @@
+    for name in names:
+        if name in coord_names:
+            found_name = name
+            break
```
there should be no problem for any sensible data

zfan001 commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,401 @@
+            found_name = name
+            break
+    if found_name is None:
+        raise ValueError(f"Unable to find a variable matching {str(names)}")
```
Done as suggested

zfan001 commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,401 @@
+    return latlon
+
+
+def get_grid_size(cube):
```
change to "get_grid_spacing"

zfan001 commented on improver/regrid/idw.py:
```diff
@@ -0,0 +1,219 @@
+WGS84_A = 6378137.0
+WGS84_IF = 298.257223563
+WGS84_F = 1.0 / WGS84_IF
+WGS84_E = math.sqrt((2.0 * WGS84_F) - (WGS84_F * WGS84_F))
```
Done as suggested

zfan001 commented on improver/regrid/idw.py:
```diff
@@ -0,0 +1,219 @@
+import numpy as np
+from scipy.spatial.ckdtree import cKDTree as KDTree
+
+WGS84_A = 6378137.0
```
added a simple comment

zfan001 commented on improver/regrid/landsea.py:
```diff
@@ -0,0 +1,454 @@
+        "bilinear-2": False,
+        "nearest-with-mask-2": True,
+        "bilinear-with-mask-2": True,
+    }
```
it is not new code. I prefer to keep it inside class

zfan001 commented on improver/regrid/landsea.py:
```diff
@@ -0,0 +1,454 @@
+                Mode of interpolation in regridding.  Valid options are "bilinear",
+                "nearest" or "nearest-with-mask".  The "nearest-with-mask" option
+                triggers adjustment of regridded points to match source points in
+                terms of land / sea type.
```
Done as suggested

zfan001 commented on improver/regrid/landsea2.py:
```diff
@@ -0,0 +1,208 @@
+        # Locate nearby input points for output points
+        indexes = basic_indexes(out_latlons, in_latlons, in_lons_dim)
+        # Initialise distances and weights to zero. Weights are only used for the bilinear case
+        distances = np.zeros((out_latlons.shape[0], 4))
```
Good point!  Done as suggested

zfan001 commented on improver/regrid/landsea2.py:
```diff
@@ -0,0 +1,208 @@
+            if BILINEAR in self.regrid_mode:
+                output_flat = apply_weights(indexes, in_values, weights)
+            else:
+                output_flat = nearest_with_mask_regrid(
```
Nice. I like it.  Done as suggested.

owena11 replied:
That structure is much more parsable, thanks!

zfan001 commented on improver/regrid/landsea2.py:
```diff
@@ -0,0 +1,208 @@
+        # Locate nearby input points for output points
+        indexes = basic_indexes(out_latlons, in_latlons, in_lons_dim)
+        # Initialise distances and weights to zero. Weights are only used for the bilinear case
+        distances = np.zeros((out_latlons.shape[0], 4))
```
added float32

zfan001 commented on improver/regrid/landsea2.py:
```diff
@@ -0,0 +1,208 @@
+        indexes = basic_indexes(out_latlons, in_latlons, in_lons_dim)
+        # Initialise distances and weights to zero. Weights are only used for the bilinear case
+        distances = np.zeros((out_latlons.shape[0], 4))
+        weights = np.zeros((out_latlons.shape[0], 4), dtype=np.float32)
```
Good point, thanks.  It will be fixed in the next commit

zfan001 replied:
I have looked into it.  np.zeros is lazy initialization, and if variables not used, memory will not be created. I tried alternative approach, and the current way seems a bit faster. Therefore, I did no change.

owena11 replied:
Huh, interesting that it's faster to quicker to do the copy, really wouldn't have thought that was that case but if it is sure!

zfan001 commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,401 @@
+         str:
+            matching name of the variable
+    """
+    coord_names = get_cube_coord_names(cube)
```
I will look into it, thanks

zfan001 replied:
Done as Aled's suggestion. just minor change "return list(matched_names)[0]".
not use iris's ....(axis="x") because we limit only lats/lons.

zfan001 replied:
Removed this function and directly use cube.coord(axis="x").standard_name as Fiona suggested.  Input cube checking will be done elsewhere.

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+    in_lons_dim,
+):
+    """
+    updating source points and weighting for one mismatched source-point cases
```
Done as suggested.

owena11 replied:
Agreed the formulation of the triangle interpolation needs further documentation that could be in a follow on PR. The current comment addresses the edge cases that result in the interpolation technique not being used, which needs to be covered.

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+            tource points's latitude-longitudes
+        out_latlons(numpy.ndarray):
+            target points's latitude-longitudes
+        in_lons_dim (int):
```
Done as suggested.

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+    n_lat = n_lat.astype(int)
+    m_lon = m_lon.astype(int)
+
+    # Four surrounding input points for each output point, in a rectangle shape
```
Nice, thanks. There were many tricky consistence issues during the development, and some code is not easily understood.  A quick ascii diagram is excellent for quick understanding.

zfan001 replied:
Added

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+
+    Returns:
+        numpy.ndarray:
+            Updated array of source grid point number for all target grid points
```
Agreed. actually thought about it before, just too time-consuming...I will try to add some

zfan001 replied:
Yes, 4 indexes of the in_latlon that bound the out_latlon point. 
Add array dimensions

zfan001 commented on improver_tests/regrid/test_RegridWithLandSeaMask.py:
```diff
@@ -0,0 +1,418 @@
+    )
+    expected_results = np.array(
+        [
+            [0.50, 0.80, 1.10, 1.40, 1.70, 2.00, 2.30, 2.60, 2.90, 3.20, 3.5],
```
Thanks. This test case is designed to check results easily for different scenarios, and nearly all types of  land-sea mask situations are covered.

zfan001 commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,401 @@
+    lats_mesh, lons_mesh = np.meshgrid(lats_data, lons_data, indexing="ij")
+    lats = lats_mesh.flatten()
+    lons = lons_mesh.flatten()
+    latlon = np.dstack((lats, lons)).squeeze()
```
This code will be replaced by the existing utility function, as Fiona suggested

zfan001 commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,401 @@
+    lats_data = cube.coord(lats_name).points
+    lons_data = cube.coord(lons_name).points
+    lats_mesh, lons_mesh = np.meshgrid(lats_data, lons_data, indexing="ij")
+    lats = lats_mesh.flatten()
```
Ordering with meshgrid/dstack/reshape is tricky and  I put efforts to make sure that coding is consistent. Yes, more notes should be added into the code .

zfan001 commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,401 @@
+    return lats_name, lons_name
+
+
+def latlon_from_cube(cube):
```
Very good point, thanks. I am a newcomer in Improver, and not familiar with utilities. I dislike rebuilding the wheel.   I will replace it using existing one. Both codes are actually similar. It seems that np.meshgrid could replace 4 lines in  transform_grid_to_lat_lon(cube). I will test it.

zfan001 replied:
We removed convert_from_projection_to_latlons, and used transform_grid_to_lat_lon (verified).

fionaRust replied:
Thank you

zfan001 commented on improver/regrid/landsea2.py:
```diff
@@ -0,0 +1,208 @@
+
+class RegridWithLandSeaMask(BasePlugin):
+    """
+    Replace data values at points where the nearest-regridding technique
```
I will change it in the next commit.

fionaRust replied:
I'm still not sure this docsting is correct

zfan001 replied:
done

zfan001 commented on improver/standardise.py:
```diff
@@ -225,305 +164,3 @@ def process(
         return cube
-
-
-class RegridLandSea(BasePlugin):
```
Tom pull regridding functions from standardise into a new created directory (regrid). It is neater.

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+        if i == 0:
+            lat2_lat = lat_1 + lat_spacing - out_lats
+            lon3_lon = lon_1 + lon_spacing - out_lons
+            weight2 = lat_spacing * lon3_lon / lat_lon_area
```
It is basically using triangle interpolation function, but we keep four weighting so that we can handle it in unified way for follow-up calculation. All interpolation formulations are explicitly given for 4 different cases. We also deferred cases if the point is not within the triangle.

fionaRust replied:
I think this will all be addressed by follow on documentation PR now so resolving thread

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+    if points_with_no_match.shape[0] > 0:
+
+        # revert back to the basic bilinear weights, indexes unchanged
+        no_match_indexes = lake_island_indexes[points_with_no_match]
```
because we make false source point's weighting zero on line 199

owena11 replied:
Ah thanks, missed that!

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+        good_count = 0
+        for i in range(k_nearest):
+            # Look for an input point with same surface type as output
+            if surface_type_mask_updates[points_with_match[point_idx], i]:
```
only new matching points has surface_type_mask "True", otherwise "False". If it is mixed with old point, it does not matter as old surface mask is "False" (not considered in follow-up calculation. 
In equal distance cases, randomly picked ... based on list order in KDtree result.

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+        weights (numpy.ndarray):
+            array of source grid point weighting for target grid points
+        in_values (numpy.ndarray):
+            input values (maybe multidimensional)
```
multidimensional case has been pre-processed using "flatten_spatial_dimensions", and then reverse back to multi-dimension using "unflatten_spatial_dimensions" (in process).

fionaRust replied:
I think putting something in the docstring to remind people of the shape of these input arrays would be useful, and the shape of what is returned. I think this is probably true for nearly all, if not all of these functions

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+    """
+    in_values_expanded = (np.ma.filled(in_values, np.nan))[indexes]
+    weighted = np.transpose(
+        np.multiply(np.transpose(weights), np.transpose(in_values_expanded))
```
All bilinear results in 3 notebooks are the same as Iris's bilinear results. Basically it confirmed that the code is correct in this regard. The ordering consistence in the code is not so straightforward. I will think about how to make it easier to read...

zfan001 commented on improver/regrid/idw.py:
```diff
@@ -0,0 +1,219 @@
+
+    out_latlons_updates = out_latlons[idw_out_indexes]
+    k_nearest = 4
+    distances_updates, indexes_updates = nearest_input_pts(
```
It is mainly for distance calculation. Initial coding used 3 types of distance calculation: vincentry distance, 'tunnel" distance and direct distance in rectilinear system. To use ecef_coords is based on a balanced consideration between speed and accuracy for IDW.

owena11 replied:
Thanks for the explanation.

zfan001 commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,401 @@
+    return out_latlons
+
+
+def classify_output_surface_type(cube_out_mask):
```
Yes, the function flattens only spatial dimensions, and other left-over dimensions are kept. It is consistent with 1-d ordering in latlons.

zfan001 replied:
classify_input_surface_type and classify_output_surface_type are used for function names for consistence, though classify_output_surface_type only requires a reshape.

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+        lon_1 = in_latlons[indexes[indexes_with_i_mismatched, 0], 1]
+
+        # Calculate updated weights
+        if i == 0:
```
use direct indexing so that the simple interpolation formulation can be explicitly set up.

zfan001 commented on improver/regrid/nearest.py:
```diff
@@ -0,0 +1,294 @@
+    """
+    min_index = np.argmin(distances, axis=1)
+    index0 = np.arange(min_index.shape[0])
+    index_in = indexes[index0, min_index]
```
I used the fancy indexing. indexes[:, min_index] results is not what I try to get here.

owena11 commented:
@zfan001 , No hurry from my side (I'm on leave next week 24th - 4th), but you've answered lots of my comments from the first pass of reviewing it. Any code changes to push up from them? So I can 'resolve conversations' and take another look at this.

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+    lat_1 = in_latlons[indexes[index_range, 0], 0]
+    lon_1 = in_latlons[indexes[index_range, 0], 1]
+    # Input point 2 - right bottom input point
+    lat2_lat = lat_1 + lat_spacing - out_lats
```
Input grid checking will be added using existing functions.

zfan001 replied:
added a function "check_if_input_grid_is_valid"

owena11 replied:
Thanks makes that assumption clearer.

zfan001 commented:
@owena11  @dmentipl 

Code modifications have been made according to your suggestions (commit 3986b5a ). 
More changes will be made once Fiona pushes up more review comments. 
Some update on acceptance test data will be updated as the master branch changed SHA256UMS.

BTW, I created a pull request for adding four jupyter notebooks in improver_aux. 

Kind regards, Julian

fionaRust commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,507 @@
+    vicinity,
+):
+    """
+    updating source points and weighting for mismatched-source-point cases
```
A lot more detail could be added to this docstring to make it easier to what calculations this function is trying to do. Even just bullet points outlining the main steps taken in the function to deal with different cases would help

zfan001 replied:
As we discussed, there are several pages of algorithm description. It is better to put somewhere else. Basically it is a science+software  project.  I reckon that we should not put science part into the code.

zfan001 replied:
Add 11-line summary of the algorithm

fionaRust replied:
Thank you

fionaRust commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,395 @@
+    return cube_in, cube_in_mask
+
+
+def create_regrid_cube(cube_array, cube_in, cube_out):
```
We have several utilities for creating new cubes within IMPROVER, but they all assume the new cube we want to create has the same x-y dimensions as one of the input cubes. I need to go back and check, and maybe chat to someone else from the Met Office, about whether any of these could be reused here to help ensure consistency.

zfan001 replied:
This function has some special requirement here, as we need cube_in for some coord  and other information.
In the new code, we convert target's curvilinear system into lat/lon first, and then go back to original system. 
Please check if the function is consistent with your requirement. BTW, BOM is going to use Albers equal-area projection, I reckon that the new code can handle it, and  I will do some testing soon.

fionaRust replied:
I've chatted this through with some people at the Met Office and we agree that this function has the correct behaviour. But I've put a couple of comments below about how it can be done neater/clearer.

fionaRust commented on improver/regrid/idw.py:
```diff
@@ -0,0 +1,218 @@
+    clons = np.cos(rlons)
+    slats = np.sin(rlats)
+    slons = np.sin(rlons)
+    n = WGS84_A / np.sqrt(1.0 - (WGS84_E * WGS84_E * slats * slats))
```
Why is the geoid hard coded here rather than taken from the input cube?

zfan001 replied:
It is actually copied from an existing code: preprocessing for calling KDtree.

fionaRust replied:
If you pass in the input cube you can access the coordinate system by doing
coord_system = cube.coord("latitude").coord_system
If it has the correct metadata I think it should return an iris,coord_systems.GeogCS instance  ( https://scitools-iris.readthedocs.io/en/stable/generated/api/iris/coord_systems.html#iris.coord_systems.GeogCS) that should have attributes you can interrogate to find all the geiod information. You can always default to WGS84 parameters if you don't find any on the cube. I'd be happy if this was done in a follow up PR based on where we are now if you want to make this change

fionaRust replied:
This will be dealt with in a follow on PR. Recorded on #1464

fionaRust commented on improver/regrid/landsea2.py:
```diff
@@ -0,0 +1,212 @@
+
+    def __init__(self, regrid_mode="bilinear-2", vicinity_radius=25000.0):
+        """
+        Initialise class
```
I think you need a docstring here to explain what the arguments are.

zfan001 replied:
added

fionaRust commented on improver_tests/regrid/test_RegridWithLandSeaMask.py:
```diff
@@ -0,0 +1,418 @@
+from iris.coords import DimCoord
+from iris.cube import Cube
+
+import improver.cli as imcli
```
I also think the code doesn't work if the input and output domain have exactly the same extent. It seems to work if the lower edges match but not the upper edges. Is this expected?

zfan001 replied:
If the input/output domains have exactly the same upper edge, it did cause problems. I have removed this limitation in the latest commit. Thank you very much for reporting the issue.

fionaRust commented on improver/regrid/landsea2.py:
```diff
@@ -0,0 +1,212 @@
+
+    def process(self, cube_in, cube_in_mask, cube_out_mask):
+        """
+        Regridding considering land_sea mask. please note cube_in must use
```
I think we should check that the input cube is in the right projection and doesn't have variable resolution. There are some helper functions we might be able to adapt for this.

zfan001 replied:
added a function "check_if_input_grid_is_valid".

fionaRust replied:
I can't see this function. Is it under the right name?

fionaRust reviewed (commented):
Here are a few more comments and questions.

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+    return out_values
+
+
+def basic_indexes(out_latlons, in_latlons, in_lons_dim):
```
I will change it for consistence.

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+    """
+    updating source points and weighting for mismatched-source-point cases
+
+    Args:
```
done

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+
+    # Initialise weights to zero at locations with mismatched surface types
+    mismatched_surface_type = np.where(np.logical_not(surface_type_mask))
+    # this would be safer with a copy to avoid values unexpectedly changing without
```
removed

zfan001 commented on improver/regrid/idw.py:
```diff
@@ -0,0 +1,219 @@
+    return x, y, z
+
+
+def similar_surface_classify(in_is_land, out_is_land, nearest_in_indexes):
```
Moved

zfan001 commented:
I have rebased the branch, and also fixed a problem with the identical upper boundary of input and output domains.
I also added one more jupyter notebook into improver_aux, in which we added more intermediate results and a grid picture. The output results can be easily verified with these information.

zfan001 commented:
@tjtg I have just made some changes on acceptance test data and sha256sums file.  "regrid part" of acceptance test passed. I will share new acceptance test data files with you.
Kind regards, Julian

zfan001 commented:
Updated to  improver_py37_iris24, and did some fixing for acceptance test.
While changing from improver_py36_iris22  to improver_py37_iris24, there is a problem for acceptance test data because the results are different under two envirnoments at two grid points for bilinear-with-mask-2 option. 
Further debugging showed that KDtree produced different results under two IMRPOVER environments. They picked different points with very similar distance, and it lead to 0.03degree difference at the final result difference at two grid points. No problem for other >1 million grid points.
 
Kind regards, Julian

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+    """
+    # Set up input points spacing values
+    lat_spacing = in_latlons[in_lons_dim, 0] - in_latlons[0, 0]
+    lon_spacing = in_latlons[1, 1] - in_latlons[0, 1]
```
It is a tricky issue. I reckon that IMPROVER handles this issue already. I checked the code, and found no code in this regard. Some existing code is just using "minus" (similar to what I am doing).

owena11 replied:
Worked through an example with Julian via teams. Seems as though I was being a bit to paranoid about distance calculation, with the assumptions in place of always accending this distance calculation is safe.

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+    out_lons = out_latlons[index_range, 1]
+    # Calculate weights for four surrounding input points
+    # Input point 1 - left bottom input point
+    lat_1 = in_latlons[indexes[index_range, 0], 0]
```
It is the normal way for finite-element approach. The text book's formulations will not order points from 0. So normally the formulation is using [1,4], and the coding follows formulation. I used to code C++ for finite element, and the same approach.

zfan001 replied:
change indexes to 1~4 for consistence

owena11 replied:
Thanks! The consistency between the two function makes it much easier to relate the two!

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+    weight2 = lat_lat1 * lon3_lon / latlon_area
+    weight3 = lat_lat1 * lon_lon1 / latlon_area
+    weight4 = lat2_lat * lon_lon1 / latlon_area
+    weights = np.transpose([weight1, weight2, weight3, weight4])
```
it depends on how you draw the figure. I used the picture in the latest figure in my fifth notebook I submittted for Improver_aux.

fionaRust commented on improver/regrid/landsea.py:
```diff
@@ -0,0 +1,454 @@
+            iris.cube.Cube: Regridded cube
+        """
+        if regrid_mode in ("nearest-with-mask-2", "bilinear-with-mask-2"):
+            if self.landmask_name not in self.landmask_source_grid.name():
```
Could still be addressed

zfan001 replied:
Done.

owena11 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,502 @@
+    # Input point 2 - right bottom input point
+    lat2_lat = lat_1 + lat_spacing - out_lats
+    # Input point 3 - right top input point
+    lon3_lon = lon_1 + lon_spacing - out_lons
```
Discussed, no upsides / downsides to either approach.

owena11 commented:
Created #1457  to track the follow on changes we're discussing on this PR so they don't get lost in the `merged and forgotten about` pile! Please edit / reword / add to as useful.

zfan001 commented:
There is a tiny spacing difference due to rounding error in Australia domain grid.  Changes have been made for considering it.
(1) improve "improver/utilities/spatial.py ", allowing tiny spacing difference  (rtol=1.e-5)
(2) pre-calculate lat_spacing/lon_spacing.
(3) change check_if_input_grid_is_valid to calculate_input_grid_spacing, and the input grid checking kept inside.

Kind regards, Julian

fionaRust commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,394 @@
+    # calculate grid spacing
+    lon_spacing = calculate_grid_spacing(cube_in, "degree", axis="x")
+    lat_spacing = calculate_grid_spacing(cube_in, "degree", axis="y")
+    print("lon_lat_spacing", lon_spacing, lat_spacing)
```
Rough print statement

zfan001 replied:
removed. Thanks

fionaRust commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,393 @@
+    Returns:
+         iris.cube.Cube: regridded result cube
+    """
+    cube_coord_names = get_cube_coord_names(cube_in)
```
I wonder if this code could be simpler if you find all your cube coordinates you want to copy across, add on the new x and y coordinates and then build your cube all at once.

Here's some half written code for you to adapt if you wish.

```  
  aux_coords_and_dims, dim_coords_and_dims = [
        [
            (coord.copy(), template_cube.coord_dims(coord))
            for coord in getattr(template_cube, coord_type) if coord not lat or lon
        ]
        for coord_type in ("aux_coords", "dim_coords")
    ]

Append new x and y coords with corresponding dims to the coord lists.

    cube = iris.cube.Cube(
        data,
        units=cube_in.units,
        attributes=cube_in.attributes,
        ...
        dim_coords_and_dims=dim_coords_and_dims,
        aux_coords_and_dims=aux_coords_and_dims,
    )
    cube.rename(name)

    return cube
```

zfan001 replied:
I will try it.

fionaRust reviewed (commented):
I have about 4 more functions I want to have a final glace over but this is the bulk of my comments based on my final readthrough

zfan001 commented on improver_tests/acceptance/test_regrid.py:
```diff
@@ -203,3 +203,91 @@ def test_regrid_nearest_landmask_multi_realization(tmp_path):
     acc.compare(output_path, kgo_path)
+
+
+def test_regrid_nearest_2_multi_realization(tmp_path):
+    """Test nearest-2 neighbour regridding"""
+    kgo_dir = acc.kgo_root() / "regrid"
+    kgo_path = kgo_dir / "nearest_2/kgo_multi_realization.nc"
+    input_path = kgo_dir / "landmask/global_cutout_multi_realization.nc"
+    target_path = kgo_dir / "landmask/ukvx_landmask.nc"
+    output_path = tmp_path / "output.nc"
+    args = [
+        input_path,
+        target_path,
+        "--output",
+        output_path,
+        "--regrid-mode",
+        "nearest-2",
+        "--regridded-title",
+        GLOBAL_UK_TITLE,
+    ]
+    run_cli(args)
+    acc.compare(output_path, kgo_path)
```
Comparing nearest-neighbour results is tricky. As I mentioned,  there are alternative points with equal distance, and it will possibly lead to different results at some points. In the five Jupyter notebooks I gave, 3/5 are the same, and 2/5 different. I debugged this kind of cases and all pointed to the points with equal distance. 

We chose a multi-realization case for avoiding the comparison, and also for more general cases.

zfan001 commented on improver_tests/acceptance/test_regrid.py:
```diff
@@ -203,3 +203,91 @@ def test_regrid_nearest_landmask_multi_realization(tmp_path):
+    acc.compare(output_path, kgo_path)
+
+
+def test_regrid_bilinear_2_multi_realization(tmp_path):
+    """Test bilinear-2 neighbour regridding"""
+    kgo_dir = acc.kgo_root() / "regrid"
+    kgo_path = kgo_dir / "bilinear_2/kgo_multi_realization.nc"
+    input_path = kgo_dir / "landmask/global_cutout_multi_realization.nc"
+    target_path = kgo_dir / "landmask/ukvx_landmask.nc"
+    output_path = tmp_path / "output.nc"
+    args = [
+        input_path,
+        target_path,
+        "--output",
+        output_path,
+        "--regrid-mode",
+        "bilinear-2",
+        "--regridded-title",
+        GLOBAL_UK_TITLE,
+    ]
+    run_cli(args)
+    acc.compare(output_path, kgo_path)
```
Bilinear regridding is always comparable. I will add this test case. Thanks.

zfan001 replied:
If I do it today, it will have impact on the current acceptance data (transfer new data). I prefer to do it after the follow-up PR. Thanks!

fionaRust replied:
Ok, put it on the follow on ticket

zfan001 commented on improver/regrid/landsea.py:
```diff
@@ -0,0 +1,412 @@
+                    cube, self.landmask_source_grid, target_grid
+                )
+
+            # identify grid-describing attributes on source cube that need updating
+            # MOSG_GRID_ATTRIBUTES={"mosg__grid_type","mosg__grid_version", "mosg__grid_domain"}
+            required_grid_attributes = [
+                attr for attr in cube.attributes if attr in MOSG_GRID_ATTRIBUTES
+            ]
+
+            # update attributes if available on target grid, otherwise remove
+            for key in required_grid_attributes:
+                if key in target_grid.attributes:
+                    cube.attributes[key] = target_grid.attributes[key]
+                else:
+                    cube.attributes.pop(key)
```
Done. Thanks

zfan001 commented on improver/regrid/landsea2.py:
```diff
@@ -0,0 +1,234 @@
+    def process(self, cube_in, cube_in_mask, cube_out_mask):
+        """
+        Regridding considering land_sea mask. please note cube_in must use
+        lats/lons rectlinear system(GeogCS). cube_in_mask and cube_in could be
```
It is what our GOCF (operational post-processing code) uses. Two files have different resolution. The current code cover both cases. Let us to discuss it during the follow-up PR.

fionaRust replied:
Yep ok, noted on #1464

zfan001 commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,393 @@
+            input cube
+    Return:
+        lat_d,lon_d (float):
+            latitude/logitude grid size
```
This function can be removed after I added calculate_input_grid_spacing last time.  Removed

zfan001 commented on improver/regrid/grid.py:
```diff
@@ -0,0 +1,393 @@
+    in_values = cube.data
+    coord_names = get_cube_coord_names(cube)
+    lats_name, lons_name = latlon_names(cube)
+    lats_index = coord_names.index(lats_name)
```
I will try it.

zfan001 replied:
it works. done.

zfan001 commented on improver/regrid/bilinear.py:
```diff
@@ -0,0 +1,635 @@
+        numpy.ndarray (total number of target points, 4):
+            Updated array of source grid point number for all target grid points
+    """
+    # find a list of target points with its latitude
```
A figure is the best explanation for this part.  considering to add something in the follow-up PR.

fionaRust replied:
Fixing in a follow on PR is fine. I think you're right a diagram would be useful.

zfan001 commented on improver/regrid/nearest.py:
```diff
@@ -0,0 +1,294 @@
+             indexes: updated array of four source grid point number for each target grid points
+ 
+    """
+    if distances.shape != surface_type_mask.shape:
```
safety check. it  never happen. Removed.

zfan001 commented:
@fionaRust Thanks for the new issues you raised. I fixed most of them, and will do more changes soon. WGS84-related change  will be considered in the follow-up PR.

zfan001 commented:
@Aled thank you for raising "Follow on Documentation for Land Sea Aware Regridding #1457" and adding additional notebooks. I will look into them after I finish the work on unit/acceptance tests.

zfan001 commented:
It is nice to see that both codecov/patch and codecov/project are improved. Only 4 lines are not run in the unit test. It is mainly due to the carefully designed unit test cases in which every scenario is turning up and every path is covered.

owena11 commented on improver_tests/regrid/test_RegridWithLandSeaMask.py:
```diff
@@ -0,0 +1,418 @@
+    np.testing.assert_allclose(regrid_nearest.data, expected_results, atol=1e-3)
+
+
+def test_regrid_bilinear_2():
```
More specific detail of the expected coordinated added in the checking of input data that makes this comment irrelevant.

fionaRust commented on improver_tests/regrid/test_RegridWithLandSeaMask.py:
```diff
@@ -0,0 +1,359 @@
+    out_mask[7, 6] = 1
+    out_mask[1, 0] = 0
+
+    cube_in = create_cube(data, in_lats, in_lons, "air_temperature", "Celsius")
```
I would still prefer you to create cubes of the right shape using the standard set up cube functions and then just replaced the points in the x and y coordinates after the cubes have been created. cube.coord("latitude).points = new_points etc.  It really is important that the cubes are set up in a standard way. I don't think this is a big change for you to make.

zfan001 replied:
I did it in the way you suggested. It works after resetting bounds. Thanks.

owena11 reviewed (approved):
Final few comments to address from @fionaRust , but PR is looking in much better shape now, thanks for addressing the review comments Julian! It's dragged on longer than is pleasant for anyone, nice to be close to the end now 👍 No more comments from me, all covered by the follow on tickets etc. 

Taken the AT data provided by Tom, unittest and acceptance tests seem to all run fine on my MO machine.

fionaRust reviewed (changes requested):
Thanks Julian. This is really close now, thank you very much for all your work on this and your effort to address all your comments. This PR is adding a lot of good functionality to IMPROVER.

Only a few comments left from my latest review. The only one that is essential to address is using the set up cube function in the unit tests.

zfan001 commented:
I fixed three issues today, and did not add the extra acceptance test. I would like to leave it to the follow-up PR to avoid the acceptance data transfer at this stage.  All cube creation in the unit test uses "set_up_variable_cube"  plus some work-around for rectangle grid. I did many tests, and it works well.

The small PR has been merged.  Two changed files in that PR are integrated into this PR as well, and so there should be no problem when merging this PR.  Please check what I did today. Hope that everything goes fine, and this PR can be merged soon. Thanks

zfan001 commented:
@Fiona  @owena11 @dmentipl @tjtg Many thanks!
